design/one-pager-crd-scaling.md

```diff
@@ -0,0 +1,150 @@
+- Establish a common understanding around relevant components and issues
+- Discuss and contribute to relevant upstream discussions with the broader community
+- Come up with tooling that can help reproducing the issues and assessing the effectiveness of considered solutions
+- Establish a common understanding on the Crossplane scenarios and if possible, a set of “definition of done” criterias for the CRD scaling issues, i.e., have a clear expectation on the scenarios we would like to support and on the expected performance in those scenarios.
+- …
+
+ 
+
+## Issues
+
+We can categorize the issues that we observe when scaling a cluster in the number of installed CRDs in two:
+- Client-side issues: Clients of the control-plane (`kubectl`, controllers, etc.) experience extended delays in the requests they make. So far, our observation is that all the issues that we categorize as client-side issues are due to “client-side throttling”, and not caused by the API server itself. These clients can throttle the requests they make to the API server solely as a function of the [rate limiter](https://pkg.go.dev/k8s.io/client-go/util/flowcontrol#RateLimiter) implementation they are using with no feedback from the server. Per our observations, the problematic client-side component that’s responsible for the high number of requests in a unit time (high request rate) is the [discovery client](https://pkg.go.dev/k8s.io/client-go/discovery#DiscoveryInterface). In the [Client-side Throttling](#client-side-throttling) section below, we will examine how `kubectl` behaves when there is a large number of CRDs installed in the cluster and how the discovery cache and the discovery client affect the perceived performance of the `kubectl` commands run. 
+- Server-side issues: ...
+
+
+### Client-side Throttling
+`kubectl` maintains a discovery cache for the discovered server-side resources under the (default) filesystem path of `$HOME/.kube/cache/discovery/<host_port>/`. Here, `<host_port>` is a string derived from the API server host and the port number it’s listening on. An example path would be `$HOME/.kube/cache/discovery/exampleaks_8e092dad.hcp.eastus.azmk8s.io_443`, or `$HOME/.kube/cache/discovery/EB788B3B801893B684B4579B2ADF0171.gr7.us_east_1.eks.amazonaws.com`. Under this cache, we have the `servergroups.json` file, which is a JSON-serialized [`v1.APIGroupList`](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#APIGroupList) object. Thus, the cache file `$HOME/.kube/cache/discovery/<host_port>/servergroups.json` holds all of the discovered API GroupVersions (GVs) together with their preferred versions from that API service. And for each discovered API GroupVersion, we have a `serverresources.json` that caches metadata about the discovered resources under that GroupVersion, JSON-serialized as a [`v1.APIResourceList`](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#APIResourceList). This metadata about resources is crucial for various tasks, such as: 
```

Comment:
I want to note that that while fixing `kubectl` would be very impactful, we may want to go down a level and try get this fixed in `client-go` so that all (Go based) clients that do discovery are fixed.

Reply:
Agreed. Cache busting for the discovery client is under discussion now. Maybe we can extend this one-pager by mentioning some options considered (cache busting, disabling client-side throttling, etc.).